### PR TITLE
Implement PixelShader in Brain.

### PR DESCRIPTION
**Pull request by xian**
- Add 24-bit RGB encoding, and 1-, 2-, and 4-bit palette lookup encodings.
- We need to know how many pixels we'll be rendering so we can allocate
  a shader buffer, so pass a Surface (including pixel count) to shaders'
  constructors.
- Wrap shader config in a Msg for easy parsing.

### CODE REVIEW

File: brain/sw/components/net/include/msg.h [hunk 1]
```diff
@@ -78,6 +78,13 @@ class Msg {
         return m;
     }
 
+    void reuse(uint8_t* buf, size_t capacity, size_t used, size_t cursor = 0) {
```
[the remark this answers is not in the document]
Comment (xian) [reply brain/sw/components/net/include/msg.h]: Yep, I was aiming for minimal changes but that was my thought too.

File: brain/sw/components/net/include/msg.h [hunk 2]
```diff
@@ -260,21 +269,24 @@ class Msg {
...
-
-    // This probably isn't useful
+    /**
+     * Copies at most `len` bytes, or at least as many bytes are available to be read.
+     *
+     * @param dest The buffer to fill.
+     * @param len The size of the `dest` buffer.
+     * @return The number of bytes read.
+     */
     size_t readBytes(uint8_t* dest, size_t len) {
```
[the remark this answers is not in the document]
Comment (xian) [reply brain/sw/components/net/include/msg.h]: The point is to get data out of the transient UDP buffer, which will presumably be returned to a pool and reused as soon as it's been handled. We can put that off until we get to the point of interframe tweening I guess.

File: brain/sw/components/shaders/compositor-shader.cpp [hunk 3]
```diff
@@ -3,41 +3,25 @@
...
-    if (*ppCursor < pEnd - 1) {
-        m_shaderA = Shader::createShaderFromDescrip(ppCursor, pEnd);
-
-        if (*ppCursor < pEnd) {
-            m_shaderB = Shader::createShaderFromDescrip(ppCursor, pEnd);
-        }
-    }
+CompositorShader::CompositorShader(Surface *surface, Msg *config) : Shader(surface) {
+    m_shaderA = Shader::createShaderFromDescrip(surface, config);
+    m_shaderB = Shader::createShaderFromDescrip(surface, config);
```
[the remark this answers is not in the document]
Comment (xian) [reply brain/sw/components/shaders/compositor-shader.cpp]: Let's discuss live; I think having zombie shaders which still consume buffer bytes is probably what we want to do.

[the remark this answers is not in the document]
Comment (xian) [reply brain/sw/components/shaders/compositor-shader.cpp]: ... the only case where you'd get a null shader is for an unknown shader type, which we could consider either catastrophic, or precede the buffers with a length so we can just skip directly.

File: brain/sw/components/shaders/pixel-shader.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    if (!pMsg->available(1)) {
+        ESP_LOGE(TAG, "invalid config");
+        disabled = 1;
+    }
+
+    m_encoding = static_cast<Encoding>(pMsg->readByte());
+    // ESP_LOGD(TAG, "pixelCount = %x", m_pixelCount);
+
+    m_dataBufSize = bufferSizeFor(m_encoding, surface->pixelCount());
+    m_dataBuf = static_cast<uint8_t *>(malloc(m_dataBufSize));
```
[the remark this answers is not in the document]
Comment (xian) [reply brain/sw/components/shaders/pixel-shader.cpp]: Let's discuss live.

File: brain/sw/components/shaders/shade-tree.cpp [hunk 5]
```diff
@@ -156,7 +156,10 @@ ShadeTree::buildNewTree() {
         // Shaders are expressed in a tree structure so we actually only need to
         // read a single root. However, that root might be a composite that
         // goes on to read further branches.
-        m_pCurrentShader = Shader::createShaderFromDescrip(&pCursor, pEnd);
+        Msg shaderConfig;
```
[the remark this answers is not in the document]
Comment (xian) [reply brain/sw/components/shaders/shade-tree.cpp]: We should never hold Msgs longer than the packet handler IMO.

[the remark this answers is not in the document]
Comment (xian) [reply brain/sw/components/shaders/shade-tree.cpp]: Pixel shader is an unusual case; most of the time we're talking about a few bytes of buffer data. Far better to separate wire protocol decoding from rendering. e.g. floats should be converted to native format and byte alignment for perf.

File: brain/sw/components/shaders/compositor-shader.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (xian) [reply brain/sw/components/shaders/compositor-shader.cpp]: I think the code is functionally equivalent, which is to say that it's probably horribly broken if we fail for any reason, but we need to improve the protocol to fix that. :-/

[the remark this answers is not in the document]
Comment (tomseago) [reply brain/sw/components/shaders/compositor-shader.cpp]: Agree with the fundamental "if it's broken it's really bad" attitude.

File: brain/sw/components/shaders/shade-tree.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomseago) [reply brain/sw/components/shaders/shade-tree.cpp]: After our in person conversations I think we're largely on the same page.

**Pull request merged**